[deliver] Implements `verify` with `altool` for Xcode 14 validation

polpielladev opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
As part of our CI, we use `deliver`'s `--verify_only` flag (implemented by myself a few months back) to validate that our binary is fit to be uploaded to the App Store every night. 

### Description
Since the removal of the Transporter app from the Xcode 14.0 toolchain, this has stopped working for us. I have seen that some changes  have already been implemented by @freddi-kit and released to migrate some of `deliver` and `pilot`'s functionalities to using `altool`.

This PR builds on top of those changes and adds missing functionality to make the `--verify_only` flag work with `altool` by using its `--validate-app` option.

### Testing Steps
1. Archive an app and export an IPA with an invalid (for upload) version of Xcode. An example of this would be Xcode 14.1 beta 2.
2. Run deliver with the `verify_only` flag: `bundle exec fastlane deliver --ipa your_awesome_app.ipa --verify_only --verbose`
3. Validation should fail with the following error
<img width="1410" alt="CleanShot 2022-10-11 at 22 54 39@2x" src="https://user-images.githubusercontent.com/24246926/195205492-794c2ba0-6ea7-436b-a693-5ee7bc0f4eac.png">

@joshdholtz @freddi-kit Feel free to give me a shout if you need extra context or if there is something you'd like me to change 👍

polpielladev committed:
Adds support for `build_verify_command`

polpielladev committed:
Updates naming of variable to show the inclusion of `verify` command

polpielladev committed:
Adds required `platform` option

polpielladev committed:
Allows passing in an `asset_path` to upload values correctly

polpielladev committed:
Updates unit tests for `deliver` component

polpielladev committed:
Fixes failing tests in `itunes_transporter_spec`

polpielladev committed:
Runs `rubocop -a`

pol-piella commented:
Hey @giginet 👋 

Did you get a chance to look over this PR? I can see that you assigned yourself to it and was wondering if you had any updates. This flag is something we use at work on a nightly workflow and has quite a lot of value for us, so I'm happy to pick up any improvements/fixes that need to be done to get the flag working again with `altool`.

polpielladev committed:
Merge branch 'fastlane:master' into master

iwc commented:
Yes, this breaks a workflow for my team as well. This would be great to have.

pol-piella commented:
@giginet Did you get a chance to take a look at this PR? Happy to go through any changes that need making or give some more context on what the changes are doing 👍

giginet commented on fastlane_core/lib/fastlane_core/itunes_transporter.rb:
```diff
@@ -47,7 +47,7 @@ def build_upload_command(username, password, source = "/tmp", provider_short_nam
     end
 
-    def build_verify_command(username, password, source = "/tmp", provider_short_name = "", jwt = nil)
+    def build_verify_command(username, password, source = "/tmp", provider_short_name = "", jwt = nil, platform = nil, api_key = nil)
```
[nits] I think it's better to use `**kwargs` in the method signatures of the abstract method.

`jwt` is only needed for ITMSTransporter, and `platform` and `api_key` are only needed for altool.
So it's not better to have all parameters on the super method for code readability.

But this point is not important.

polpielladev replied:
Just modified the abstract method to use `**kwargs` 👍

giginet commented on fastlane_core/lib/fastlane_core/itunes_transporter.rb:
```diff
@@ -800,10 +812,20 @@ def upload(app_id = nil, dir = nil, package_path: nil, asset_path: nil, platform
+      actual_dir = if can_use_asset_path && !force_itmsp
+                     # The asset gets deleted upon completion so copying to a temp directory
+                     # (with randomized filename, for multibyte-mixed filename upload fails)
+                     new_file_name = "#{SecureRandom.uuid}#{File.extname(asset_path)}"
+                     tmp_asset_path = File.join(Dir.tmpdir, new_file_name)
+                     FileUtils.cp(asset_path, tmp_asset_path)
+                     tmp_asset_path
```
You can use `Tempfile.create` to randomize temporary files. https://docs.ruby-lang.org/en/3.2/Tempfile.html

polpielladev replied:
Ended up not changing the code for this as I copied it from the `upload` command. As it needs to copy a file to a directory, I couldn't find any easy ways to use `Tempfile` other than having to open and copy the contents of the asset path to the temp path. 

Are you happy with me leaving it as is or do you know of anyway you can get a temp path without creating a path? 🤔

giginet replied:
I thought again. it's a good situation to use `Dir.mktmpdir` instead. not `Tempfile.create` 🙇 
https://ruby-doc.org/stdlib-2.5.1/libdoc/tmpdir/rdoc/Dir.html

I added suggestions to use this method.
Thanks to this method, we can get rid of the randomize logic.

This tempdir is not automatically deleted, however, it will be deleted on the below line.
https://github.com/fastlane/fastlane/pull/20738/files#diff-e4392686971e0801abe75321cc8a4c90eb2f2c971734d8cc88b0ac5fef5ee12dR860

giginet commented on fastlane_core/lib/fastlane_core/itunes_transporter.rb:
```diff
@@ -812,8 +834,15 @@ def verify(app_id = nil, dir = nil, package_path: nil)
+      use_api_key = !@api_key.nil?
+      api_key_placeholder = use_api_key ? { key_id: "YourKeyID", issuer_id: "YourIssuerID", key_dir: "YourTmpP8KeyDir" } : nil
+
+      api_key = nil
```
It's unnecessary

```suggestion
```

polpielladev replied:
Removed 👍

giginet commented on deliver/lib/deliver/runner.rb:
```diff
@@ -266,7 +266,7 @@ def submit_for_review
     # If there are multiple teams, infer the provider from the selected team name.
     # If there are fewer than two teams, don't infer the provider.
-    def transporter_for_selected_team(upload: false)
+    def transporter_for_selected_team(altool_compatible_command: false)
```
What's the situation `altool_compatible_command` to be `false`?

I think this flag is not necessary for `ItunesTransporter`.

In the previous implementation, `verify` task doesn't use `altool`, so some flags are needed to decide which executors are used.
However, in the current implementation, the transporter always uses `altool` when meeting the situations.
So I think this is not needed. What do you think?

giginet replied:
I meant all callers of this method seem to pass `true` for `altool_compatible_command`

https://github.com/fastlane/fastlane/pull/20738/files#diff-51e845b40462cd52020ab428469c85f65e24e1343e0b17c88d0c1a11e1ec22ecR172
https://github.com/fastlane/fastlane/pull/20738/files#diff-51e845b40462cd52020ab428469c85f65e24e1343e0b17c88d0c1a11e1ec22ecR209

So we can get rid of this argument. We can always pass `true` on L284 to reduce unnecessary states.

giginet reviewed (commented):
@pol-piella 

Hi! I apologize to keep waiting for you for such a long time.

I noticed `altool_compatible_command` might be removed in this implementation.
What do you think? 
https://github.com/fastlane/fastlane/pull/20738/files#r1070544411

If you can fix this, could you try fixing it?

pol-piella commented:
@giginet Also, I have another very small PR to add support for a `xcodebuild` flag that would help us get some metrics at work, would you mind reviewing it if you get a chance? 

Not sure what the process is to get someone to look at a PR, is there maybe a slack channel I can send a message on?

polpielladev committed:
Merge branch 'fastlane:master' into master

Commit:
Uses `**kwargs` for abstract `build_verify_command` method

Commit:
Fixes `rubocop` violation

polpielladev commented:
@giginet I think I have addressed all comments now, sorry it took a long time to do so, been a busy few weeks 😅

Let me know what you think about them and if you have any further comments/suggestions, happy to push further updates 👍

giginet commented:
Thank you for your effort. I'll review it this weekend. 👍

giginet commented on deliver/lib/deliver/runner.rb:
```diff
@@ -281,20 +281,20 @@ def transporter_for_selected_team(upload: false)
       unless api_token.nil?
         api_token.refresh! if api_token.expired?
-        return FastlaneCore::ItunesTransporter.new(nil, nil, false, nil, api_token.text, upload: upload, api_key: api_key)
+        return FastlaneCore::ItunesTransporter.new(nil, nil, false, nil, api_token.text, altool_compatible_command: altool_compatible_command, api_key: api_key)
```
We can always pass `altool_compatible_command: true` because of the above comment.

giginet commented on fastlane_core/lib/fastlane_core/itunes_transporter.rb:
```diff
@@ -800,10 +816,20 @@ def upload(app_id = nil, dir = nil, package_path: nil, asset_path: nil, platform
 
-      actual_dir = if package_path
+      actual_dir = if can_use_asset_path && !force_itmsp
+                     # The asset gets deleted upon completion so copying to a temp directory
+                     # (with randomized filename, for multibyte-mixed filename upload fails)
+                     new_file_name = "#{SecureRandom.uuid}#{File.extname(asset_path)}"
+                     tmp_asset_path = File.join(Dir.tmpdir, new_file_name)
+                     FileUtils.cp(asset_path, tmp_asset_path)
+                     tmp_asset_path
```
```suggestion
                     # The asset gets deleted upon completion so copying to a temp directory
                     tmp_asset_path = Dir.mktmpdir
                     FileUtils.cp(asset_path, tmp_asset_path)
                     tmp_asset_path
```

giginet commented on fastlane_core/lib/fastlane_core/itunes_transporter.rb:
```diff
@@ -812,8 +838,14 @@ def verify(app_id = nil, dir = nil, package_path: nil)
-      UI.verbose(@transporter_executor.build_verify_command(@user, password_placeholder, actual_dir, @provider_short_name, jwt_placeholder))
+      # Handle AppStore Connect API
+      use_api_key = !@api_key.nil?
+      api_key_placeholder = use_api_key ? { key_id: "YourKeyID", issuer_id: "YourIssuerID", key_dir: "YourTmpP8KeyDir" } : nil
```
It's better to be commented for code readability 👍 

```suggestion
      # Masking credentials for verbose outputs
      api_key_placeholder = use_api_key ? { key_id: "YourKeyID", issuer_id: "YourIssuerID", key_dir: "YourTmpP8KeyDir" } : nil
```

giginet reviewed (commented):
It looks mostly good 👍 
I appreciate your contribution and improving code quality.

As I commented, I believe in removing the argument of `transporter_for_selected_team`.

If you can fix the comments, I'll merge this soon.

```diff
diff --git a/deliver/lib/deliver/runner.rb b/deliver/lib/deliver/runner.rb
index a2b433efe..2494af5a3 100644
--- a/deliver/lib/deliver/runner.rb
+++ b/deliver/lib/deliver/runner.rb
@@ -169,7 +169,7 @@ module Deliver
       pkg_path = options[:pkg]
 
       platform = options[:platform]
-      transporter = transporter_for_selected_team(altool_compatible_command: true)
+      transporter = transporter_for_selected_team
 
       case platform
       when "ios", "appletvos"
@@ -206,7 +206,7 @@ module Deliver
       pkg_path = options[:pkg]
 
       platform = options[:platform]
-      transporter = transporter_for_selected_team(altool_compatible_command: true)
+      transporter = transporter_for_selected_team
 
       case platform
       when "ios", "appletvos"
@@ -266,7 +266,7 @@ module Deliver
     # If itc_provider was explicitly specified, use it.
     # If there are multiple teams, infer the provider from the selected team name.
     # If there are fewer than two teams, don't infer the provider.
-    def transporter_for_selected_team(altool_compatible_command: false)
+    def transporter_for_selected_team
       # Use JWT auth
       api_token = Spaceship::ConnectAPI.token
       api_key = if options[:api_key].nil? && !api_token.nil?
@@ -281,12 +281,12 @@ module Deliver
 
       unless api_token.nil?
         api_token.refresh! if api_token.expired?
-        return FastlaneCore::ItunesTransporter.new(nil, nil, false, nil, api_token.text, altool_compatible_command: altool_compatible_command, api_key: api_key)
+        return FastlaneCore::ItunesTransporter.new(nil, nil, false, nil, api_token.text, altool_compatible_command: true, api_key: api_key)
       end
 
       tunes_client = Spaceship::ConnectAPI.client.tunes_client
 
-      generic_transporter = FastlaneCore::ItunesTransporter.new(options[:username], nil, false, options[:itc_provider], altool_compatible_command: altool_compatible_command, api_key: api_key)
+      generic_transporter = FastlaneCore::ItunesTransporter.new(options[:username], nil, false, options[:itc_provider], altool_compatible_command: true, api_key: api_key)
       return generic_transporter unless options[:itc_provider].nil? && tunes_client.teams.count > 1
 
       begin
@@ -294,7 +294,7 @@ module Deliver
         name = team['name']
         provider_id = generic_transporter.provider_ids[name]
         UI.verbose("Inferred provider id #{provider_id} for team #{name}.")
-        return FastlaneCore::ItunesTransporter.new(options[:username], nil, false, provider_id, altool_compatible_command: altool_compatible_command, api_key: api_key)
+        return FastlaneCore::ItunesTransporter.new(options[:username], nil, false, provider_id, altool_compatible_command: true, api_key: api_key)
       rescue => ex
         UI.verbose("Couldn't infer a provider short name for team with id #{tunes_client.team_id} automatically: #{ex}. Proceeding without provider short name.")
         return generic_transporter
```

polpielladev committed:
Removes `altool_compatible_command` argument from `deliver` runner

polpielladev committed:
Fixes further PR comments

polpielladev commented:
Hi @giginet 👋 Thank you so much for the help and the awesome feedback. I have addressed the comments you suggested now 👍

giginet commented on fastlane_core/lib/fastlane_core/itunes_transporter.rb:
```diff
@@ -800,10 +816,19 @@ def upload(app_id = nil, dir = nil, package_path: nil, asset_path: nil, platform
+      actual_dir = if can_use_asset_path && !force_itmsp
+                     # The asset gets deleted upon completion so copying to a temp directory
+                     # (with randomized filename, for multibyte-mixed filename upload fails)
+                     tmp_asset_path = Dir.mktmpdir
+                     FileUtils.cp(asset_path, tmp_asset_path)
+                     tmp_asset_path
```
I'm so sorry, my suggestion was completely wrong.

Filenames are not randomized. and `tmp_asset_path` indicates the parent directory path...

It have to revert to your implementation 🙏 

```suggestion
                     new_file_name = "#{SecureRandom.uuid}#{File.extname(asset_path)}"
                     tmp_asset_path = File.join(Dir.tmpdir, new_file_name)
                     FileUtils.cp(asset_path, tmp_asset_path)
                     tmp_asset_path
```

polpielladev replied:
Ah no problem! I have reverted the changes to the original implementation now 👍

polpielladev committed:
Reverts changes to temporary path code

giginet reviewed (approved):
LGTM 🚀 
  I'm sorry to wait for you to start reviewing for a long time.
  I'll merge this after checking.
  
  Thanks a lot again! 😄

polpielladev commented:
> LGTM 🚀 I'm sorry to wait for you to start reviewing for a long time. I'll merge this after checking.
> 
> Thanks a lot again! 😄

No problem at all honestly! Thanks so much for all your help on this, I really appreciate it 🎉

giginet commented:
I tried this command for the latest Xcode, it works!

```ruby
deliver(
  verify_only: true
)
```
